src/index.tsx

```diff
@@ -137,13 +155,101 @@ export class Terminal extends Component<TerminalProps, TerminalState> {
       }
     };
 
+    const handleTabPress = async (): Promise<void> => {
+      const {
+        autoCompleteActiveItem,
+        autoCompleteIsActive,
+        autoCompleteItems,
+        inputValue,
+        currentPath,
+        fileSystem,
+      } = this.state;
+      const { commandName, commandTargets } = parseCommand(inputValue);
+
+      let newAutoCompleteActiveItem = 0;
+      if (autoCompleteIsActive && autoCompleteItems) {
+        const previewItemCount = Object.keys(autoCompleteItems).length;
+
+        if (autoCompleteActiveItem < previewItemCount - 1) {
+          newAutoCompleteActiveItem = autoCompleteActiveItem + 1;
+        }
+
+        let newInputValue = `${inputValue}${
+          Object.keys(autoCompleteItems)[newAutoCompleteActiveItem]
+        }`;
+
+        if (commandTargets[0].length > 0) {
+          newInputValue = inputValue.replace(
+            commandTargets[0],
+            Object.keys(autoCompleteItems)[newAutoCompleteActiveItem],
+          );
+        }
+
+        this.setState(
+          Object.assign({
+            autoCompleteActiveItem: newAutoCompleteActiveItem,
+            inputValue: newInputValue,
+          }),
+        );
+
+        return;
+      }
+
+      let commandResult: AutoCompleteResponse['commandResult'];
+      if (commands[`${commandName}AutoComplete`]) {
+        ({ commandResult } = await commands[`${commandName}AutoComplete`](
+          fileSystem,
+          currentPath,
+          commandTargets[0],
+        ));
+      } else {
+        // Do nothing if tab is not supported
+        return;
+      }
+
+      if (commandResult) {
+        if (Object.keys(commandResult).length === 1) {
+          this.setState(
+            Object.assign({
+              inputValue: inputValue.replace(
+                commandTargets[0],
+                Object.keys(commandResult)[0],
+              ),
+            }),
+          );
+        } else {
+          this.setState(
+            Object.assign({
+              autoCompleteIsActive: true,
+              autoCompleteItems: commandResult,
+            }),
+          );
+        }
+      }
+    };
+
+    // If we do anything other than tab, clear autocomplete state
+    if (!(event.keyCode == 9 || event.key === 'Tab')) {
```

Comment:
Is there weirdness with having multiple setState calls like this? Maybe refactor